src/well_known_messages.cc

```diff
@@ -0,0 +1,248 @@
+#include <well_known_messages.h>
+
+namespace grpc_labview
+{
+    namespace wellknown
+    {
+        //---------------------------------------------------------------------
+        //---------------------------------------------------------------------
+        MetadataOwner* MetadataOwner::_instance = nullptr;
+        std::mutex MetadataOwner::_mutex;
+
+        //---------------------------------------------------------------------
+        //---------------------------------------------------------------------
+        MetadataOwner& MetadataOwner::GetInstance()
+        {
+            std::unique_lock<std::mutex> lock(_mutex);
+            if (_instance == nullptr)
+            {
+                _instance = new MetadataOwner();
+                auto owner = _instance->_owner.get();
+                owner->RegisterMetadata(Double2DArray::GetInstance().GetMetadata(owner));
+                owner->RegisterMetadata(String2DArray::GetInstance().GetMetadata(owner));
+                owner->FinalizeMetadata();
+            }
+            return *_instance;
+        }
+
+        //---------------------------------------------------------------------
+        //---------------------------------------------------------------------
+        MetadataOwner::MetadataOwner()
+        {
+            _owner = std::make_unique<MessageElementMetadataOwner>();
+        }
+
+        //---------------------------------------------------------------------
+        //---------------------------------------------------------------------
+        std::shared_ptr<MessageMetadata> MetadataOwner::FindMetadata(const std::string& name)
+        {
+            return _owner->FindLocalMetadata(name);
+        }
+
+        //---------------------------------------------------------------------
+        //---------------------------------------------------------------------
+        template<typename TRepeatedType>
+        std::shared_ptr<MessageMetadata> Base2DArray<TRepeatedType>::GetMetadata(IMessageElementMetadataOwner* metadataOwner)
```

Comment:
FYI: defining template methods in the .cpp file only works if the template is instantiated from the .cpp file. In this case, it happens to work because the same .cpp file defines a subclass for each template instance.

I guess this is fine, but if it breaks, the solution is to either move the code to the .h/.ipp file or add explicit instantiation to the .cpp file.

Reply:
I moved the implementation to the header.